[FEATURE] Support default Databricks notebook authentication

**Problem Statement**

Hello, it automatically reads the generated personal access token on the databricks notebooks for authentication without requiring separate authentication when using the Python SDK.

Simply declaring a WorkspaceClient is enough to correctly receive the authentication of the entity running the notebook, whether it's run by a user or a service principal through a workflow. 

```
from databricks.sdk import WorkspaceClient

w = WorkspaceClient()

for c in w.clusters.list():
  print(c.cluster_name)
```

I wish a similar feature was available in Scala, but it seems like the Databricks Java SDK does not support this. 

Is there any plan to support this?

**Additional Context**

https://docs.databricks.com/en/dev-tools/sdk-python.html#use-the-databricks-sdk-for-python-from-a-databricks-notebook

Oh.. I just tried declaring empty config and using it, but it didn't work well, so I didn't look deeper into the code. 

But after seeing the code you attached and actually testing it, I found that setting the setHost directly in the notebook allows calls to be made without any separate credential settings. 

However, while there's no need to set the host when executing in the notebook with the Python SDK, it seems that's not the case with Java SDK.

It seems like it should work when looking at [here](https://github.com/databricks/databricks-sdk-java/blob/7820506311047ba1b13533aac3df523a0be1bc23/databricks-sdk-java/src/main/java/com/databricks/sdk/core/NotebookNativeCredentialsProvider.java#L113-L115)

```
# 
import com.databricks.sdk.service.compute.ListClustersRequest
import com.databricks.sdk.WorkspaceClient
import com.databricks.sdk.core.DatabricksConfig

// It works in databricks notebook!
val config = new DatabricksConfig()
    .setHost("https://krafton-hq.cloud.databricks.com")
val w = new WorkspaceClient(config)
w.clusters().list(new ListClustersRequest()).forEach{
  elem => println(elem.getClusterName)
}

// It doesn't works..
// Caused by: ClientProtocolException: 
// Caused by: ProtocolException: Target host is not specified
val config2 = new DatabricksConfig()
val w2 = new WorkspaceClient(config2)
w2.clusters().list(new ListClustersRequest()).forEach{
  elem => println(elem.getClusterName)
}
```

--- a/databricks-sdk-java/src/main/java/com/databricks/sdk/core/ApiClient.java
+++ b/databricks-sdk-java/src/main/java/com/databricks/sdk/core/ApiClient.java
@@ -231,11 +231,10 @@ public class ApiClient {
   }
 
   private Response getResponse(Request in) {
-    in.withUrl(config.getHost() + in.getUrl());
-    return executeInner(in);
+    return executeInner(in, in.getUrl());
   }
 
-  private Response executeInner(Request in) {
+  private Response executeInner(Request in, String path) {
     RetryStrategy retryStrategy = retryStrategyPicker.getRetryStrategy(in);
     int attemptNumber = 0;
     while (true) {
@@ -246,6 +245,10 @@ public class ApiClient {
 
       // Authenticate the request. Failures should not be retried.
       in.withHeaders(config.authenticate());
+
+      // Prepend host to URL only after config.authenticate().
+      // This call may configure the host (e.g. in case of notebook native auth).
+      in.withUrl(config.getHost() + path);
 
       // Set User-Agent with auth type info, which is available only
       // after the first invocation to config.authenticate()

--- a/databricks-sdk-java/src/test/java/com/databricks/sdk/core/ApiClientTest.java
+++ b/databricks-sdk-java/src/test/java/com/databricks/sdk/core/ApiClientTest.java
@@ -42,26 +42,28 @@ public class ApiClientTest {
     }
   }
 
-  private ApiClient getApiClient(Request request, List<ResponseProvider> responses) {
+  private ApiClient getApiClient(DatabricksConfig config, Request request, List<ResponseProvider> responses) {
     DummyHttpClient hc = new DummyHttpClient();
     for (ResponseProvider response : responses) {
       hc.with(request, response);
     }
+    return new ApiClient(config.setHttpClient(hc), new FakeTimer());
+  }
+
+  private ApiClient getApiClient(Request request, List<ResponseProvider> responses) {
     String host = request.getUri().getScheme() + "://" + request.getUri().getHost();
     DatabricksConfig config =
-        new DatabricksConfig()
-            .setHttpClient(hc)
-            .setHost(host)
-            .setCredentialsProvider(new DummyCredentialsProvider());
-    return new ApiClient(config, new FakeTimer());
+            new DatabricksConfig()
+                    .setHost(host)
+                    .setCredentialsProvider(new DummyCredentialsProvider());
+    return getApiClient(config, request, responses);
   }
 
   private <T> void runApiClientTest(
-      Request request,
-      List<ResponseProvider> responses,
-      Class<? extends T> clazz,
-      T expectedResponse) {
-    ApiClient client = getApiClient(request, responses);
+          ApiClient client,
+          Request request,
+          Class<? extends T> clazz,
+          T expectedResponse) {
     T response;
     if (request.getMethod().equals(Request.GET)) {
       response = client.GET(request.getUri().getPath(), clazz, Collections.emptyMap());
@@ -71,6 +73,15 @@ public class ApiClientTest {
       throw new IllegalArgumentException("Unsupported method: " + request.getMethod());
     }
     assertEquals(response, expectedResponse);
+  }
+
+  private <T> void runApiClientTest(
+          Request request,
+          List<ResponseProvider> responses,
+          Class<? extends T> clazz,
+          T expectedResponse) {
+    ApiClient client = getApiClient(request, responses);
+    runApiClientTest(client, request, clazz, expectedResponse);
   }
 
   private void runFailingApiClientTest(
@@ -345,6 +356,40 @@ public class ApiClientTest {
             new Failure(new UnknownHostException("Connect timed out")), getSuccessResponse(req)),
         MyEndpointResponse.class,
         new MyEndpointResponse().setKey("value"));
+  }
+
+  class HostPopulatingCredentialsProvider implements CredentialsProvider {
+    private final String host;
+    private final CredentialsProvider parent;
+
+    public HostPopulatingCredentialsProvider(String host) {
+      this.host = host;
+      this.parent = new DummyCredentialsProvider();
+    }
+
+    @Override
+    public String authType() {
+      return parent.authType();
+    }
+
+    @Override
+    public HeaderFactory configure(DatabricksConfig config) {
+      config.setHost(this.host);
+      return parent.configure(config);
+    }
+  }
+
+  @Test
+  void populateHostFromCredentialProvider() {
+    Request req = getBasicRequest();
+    DatabricksConfig config = new DatabricksConfig()
+            .setCredentialsProvider(new HostPopulatingCredentialsProvider("http://my.host"));
+    ApiClient client = getApiClient(config, req, Collections.singletonList(getSuccessResponse(req)));
+    runApiClientTest(
+            client,
+            req,
+            MyEndpointResponse.class,
+            new MyEndpointResponse().setKey("value"));
   }
 
   @Test

--- a/databricks-sdk-java/src/test/java/com/databricks/sdk/core/ApiClientTest.java
+++ b/databricks-sdk-java/src/test/java/com/databricks/sdk/core/ApiClientTest.java
@@ -42,7 +42,8 @@ public class ApiClientTest {
     }
   }
 
-  private ApiClient getApiClient(DatabricksConfig config, Request request, List<ResponseProvider> responses) {
+  private ApiClient getApiClient(
+      DatabricksConfig config, Request request, List<ResponseProvider> responses) {
     DummyHttpClient hc = new DummyHttpClient();
     for (ResponseProvider response : responses) {
       hc.with(request, response);
@@ -53,17 +54,12 @@ public class ApiClientTest {
   private ApiClient getApiClient(Request request, List<ResponseProvider> responses) {
     String host = request.getUri().getScheme() + "://" + request.getUri().getHost();
     DatabricksConfig config =
-            new DatabricksConfig()
-                    .setHost(host)
-                    .setCredentialsProvider(new DummyCredentialsProvider());
+        new DatabricksConfig().setHost(host).setCredentialsProvider(new DummyCredentialsProvider());
     return getApiClient(config, request, responses);
   }
 
   private <T> void runApiClientTest(
-          ApiClient client,
-          Request request,
-          Class<? extends T> clazz,
-          T expectedResponse) {
+      ApiClient client, Request request, Class<? extends T> clazz, T expectedResponse) {
     T response;
     if (request.getMethod().equals(Request.GET)) {
       response = client.GET(request.getUri().getPath(), clazz, Collections.emptyMap());
@@ -76,10 +72,10 @@ public class ApiClientTest {
   }
 
   private <T> void runApiClientTest(
-          Request request,
-          List<ResponseProvider> responses,
-          Class<? extends T> clazz,
-          T expectedResponse) {
+      Request request,
+      List<ResponseProvider> responses,
+      Class<? extends T> clazz,
+      T expectedResponse) {
     ApiClient client = getApiClient(request, responses);
     runApiClientTest(client, request, clazz, expectedResponse);
   }
@@ -382,14 +378,13 @@ public class ApiClientTest {
   @Test
   void populateHostFromCredentialProvider() {
     Request req = getBasicRequest();
-    DatabricksConfig config = new DatabricksConfig()
+    DatabricksConfig config =
+        new DatabricksConfig()
             .setCredentialsProvider(new HostPopulatingCredentialsProvider("http://my.host"));
-    ApiClient client = getApiClient(config, req, Collections.singletonList(getSuccessResponse(req)));
+    ApiClient client =
+        getApiClient(config, req, Collections.singletonList(getSuccessResponse(req)));
     runApiClientTest(
-            client,
-            req,
-            MyEndpointResponse.class,
-            new MyEndpointResponse().setKey("value"));
+        client, req, MyEndpointResponse.class, new MyEndpointResponse().setKey("value"));
   }
 
   @Test

